recipe-client-addon/lib/StudyStorage.jsm

```diff
@@ -0,0 +1,210 @@
+    // otherwise we can assume it is a primitive type (like "boolean" or "number").
+    const validates = typeof requiredType === "function" ?
+      incomingData instanceof requiredType : typeof incomingData === requiredType;
+
+    if (!validates) {
+      throw new Error(`Key "${dataKey}" must be of type "${requiredType.name || requiredType}"`);
+    }
+  });
+}
+
+
+// Creates new object store and populates that store's indexes.
+function setupDatabase(db) {
+  const {schema} = DB_OPTIONS;
+  const store = db.createObjectStore(DB_NAME, { keyPath: "name" });
+
+  Object.keys(schema).forEach((schemaKey) => {
+    store.createIndex(schemaKey, schemaKey, schema[schemaKey]);
+  });
+}
+
+// Determines if a DB connection/request already exists.
+function hasDatabaseRequest() {
+  return !!gDatabase;
+}
+
+// Returns the currently active IDB connection, and creates one if not
+// already present.
+async function getDatabase() {
+  if (!hasDatabaseRequest()) {
+    // gDB is a Promise to prevent multiple calls to `IDB.open` - subsequent calls
+    // will refer to this same promise, requiring only one `open` call internally.
+    gDatabase = new Promise(async (resolve) => {
+      const db = await IndexedDB.open(DB_NAME, DB_OPTIONS, setupDatabase);
+      resolve(db);
+    });
+  }
+
+  return gDatabase;
+}
+
+
+this.StudyStorage = {
+  /**
+   * Clears all data from the connected IndexedDB.
+   * @async
+   */
+  async clearAllData() {
+    const db = await getDatabase();
+    await db.objectStore(DB_NAME, "readwrite").clear();
+  },
+
+  /**
+   * If a database exists and is connected, disconnect and remove the reference
+   * to that database.
+   *
+   * @async
+   */
+  async closeConnection() {
+    if (hasDatabaseRequest()) {
+      const db = await getDatabase();
+      await db.close();
+      gDatabase = null;
+    }
+  },
+
+  /**
+   * Look up a Study based on the given unique name. Returns `null` if no study
+   * is found in the DB.
+   *
+   * @async
+   * @param  {string} studyName Unique name of study to retrieve.
+   * @return {Promise<Study>}
+   */
+  async getStudy(studyName) {
+    const db = await getDatabase();
+
+    const foundStudy = await db.objectStore(DB_NAME, "readonly").get(studyName);
+
+    return foundStudy || null;
+  },
+
+  /**
+   * Given a Study, validates its schema and saves the data into the DB's
+   * objectStore instance.
+   *
+   * @async
+   * @param  {Study}  study [description]
+   * @return {Promise}
+   */
+  async saveStudy(study) {
+    const {
+      addonId, // The add-on ID
+      name, // The unique name for the study
+      version, // The add-on version
+      description, // The add-on description
+      studyStartDate = Date.now(), // The study start date
+      studyEndDate = null, // The study end date (starts off null)
+      active = true, // The active status of the study (starts off true)
+    } = study;
+
+    const db = await getDatabase();
```

Comment:
Would it make sense to wait for the db ref after the validation step, or is there some reason that's not possible? Then the validation results would return sync.